### PR TITLE
NIFI-4827 Added support for reading queries from the flowfile body to…

### PR DESCRIPTION
**Pull request by MikeThomsen**
… GetMongo.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml [hunk 1]
```diff
@@ -0,0 +1,635 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml]: It's a miniature test flow that's meant to show the behavior if someone has questions. I got the idea when I was reviewing the PutInfluxDb processor. On nifi-dev, I think @joewitt mentioned that a major problem with some of these bundles is the people who actually know a particular system might be highly limited in the community. I think getting into a habit of sharing test flows that use something like GenerateFlowFile could ease the learning curve on others. Particularly if a docker-compose file or something like that is added with it.

I can remove it, but I think it might be a good practice.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 2]
```diff
@@ -81,6 +82,15 @@ public ValidationResult validate(final String subject, final String value, final
     };
 
     static final Relationship REL_SUCCESS = new Relationship.Builder().name("success").description("All files are routed to success").build();
+    static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("All input flowfiles that are part of a failed query execution go here.")
```
[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: So you're saying it should just be more explicit in the documentation, and not be some sort of dynamic relationship a la RouteOnAttribute, right?

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 3]
```diff
@@ -89,6 +99,17 @@ public ValidationResult validate(final String subject, final String value, final
         .expressionLanguageSupported(true)
         .addValidator(DOCUMENT_VALIDATOR)
         .build();
+
+    static final AllowableValue LOC_BODY = new AllowableValue("body", "Body");
+    static final AllowableValue LOC_PARAM = new AllowableValue("param", "Query Parameter");
+    static final PropertyDescriptor QUERY_LOC = new PropertyDescriptor.Builder()
```
[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Everything is tied to that parameter (or should be tied to it). One of the things I'm learning from having to train someone on NiFi is that "explicit is better" for a lot of people. I think the meaning of a blank query wouldn't be obvious to a lot of users, and you cannot trust that they'll read the description closely rather than cursing out the product because it seems to be "mysteriously misbehaving."

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Correct, just thinking of other processors where this happens, their doc says something like what I put above, with no incoming connections the relationship is unused.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: We can ask the users list to get UX and other feedback, for ExecuteSQL it was natural because the query used to be required, then we relaxed it and added doc to the property saying if left blank, the contents of the flow file are expected to have the SQL to execute. In this case, a blank setting is already meaningful, so we may need two properties. "Explicit is better" is a good rule of thumb, but now we have two properties that interact with each other in different ways, which can also be confusing (see my other comments on when things are set to various value combos)

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml]: I like the idea of having test flows, but maybe not in the codebase since they aren't executed? I often put test flows up as a Gist on GitHub and refer to it in the PR, so the reviewer can go there for a sample flow (which I try to document inline with labels to explain what's going on)

[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml]: If not there, it should be some place that is owned by the NiFi team so it can easily survive us if the infamous "hit by a bus" scenario happens.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Actually, I think allowing a blank field was the wrong decision because it is not a valid query w/ MongoDB. So I'll make that change.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml]: Agreed, I often make the Gist available for the reviewer, then once merged, the template could be added to the Wiki of sample templates for all to have

[the remark this answers is not in the document]
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/test-flows/NIFI_4827.xml]: Ok. Removed this for now since it can go on the Wiki.

**Comment by MikeThomsen** [comment]
Ok. I just pushed some changes that should fix things with the query location issue. I took your advice on following the SQL approach. Unit tests are there and batter the heck out of it.

**Comment by MikeThomsen** [comment]
@mattyb149 Any chance you can take a look?

**Comment by mattyb149** [comment]
Merging #2180 caused a conflict on this PR, mind rebasing against the latest master? Please and thanks!

**Comment by MikeThomsen** [comment]
NP. I'll try to get that done in a little bit.

**Comment by MikeThomsen** [comment]
@mattyb149 Rebased and pushed. Building now.

**Comment by MikeThomsen** [comment]
@mattyb149 Passed the test and it says it's ready.

**Comment by MikeThomsen** [comment]
@mattyb149 I have time today that if you can get to this I can switch over to 4838 and get that ready for close out too.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 4]
```diff
@@ -226,6 +242,11 @@ private ObjectWriter getObjectWriter(ObjectMapper mapper, String ppSetting) {
 
     @Override
     public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
+        if (!context.hasIncomingConnection() && (context.getProperty(QUERY) == null)) {
+            throw new RuntimeException("Without an incoming connection, the Query property must be set.");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: The RuntimeException here and on line 276 should be ProcessException, it's a NiFi-specific runtime exception that indicates we realize something went wrong but we don't know how to handle any input/output we might have (like routing to failure).  Also rather than `context.getProperty(QUERY) == null` since there will be a PropertyValue object returned, you should use `context.getProperty(QUERY).isSet()`.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java [hunk 5]
```diff
@@ -69,6 +70,7 @@
     @Before
     public void setup() {
         runner = TestRunners.newTestRunner(GetMongo.class);
+        runner.setProperty(GetMongo.QUERY, "{}");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: Does this need to be here? Seems like it might hide the cases when Query is empty, which would be the existing behavior. If it's because of the validator, then it should accept an empty query as valid since the property is not required.  Also, can you add a unit test (or update the existing one) for when the Query is empty and there is no incoming flow file (but there is an incoming connection)? That would test the other half of the conditions for the exception block you added at the top of onTrigger.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java [hunk 6]
```diff
@@ -296,4 +298,55 @@ public void testQueryAttribute() {
...
+        boolean exception = false;
+        try {
+            runner.assertValid();
+            runner.setIncomingConnection(false);
+            runner.run(1, true, true);
+        } catch (Throwable pe) {
+            exception = true;
+        }
+
+        Assert.assertTrue("No exception was thrown!", exception);
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: After changing to throw a ProcessException, you could instead have an Exception object set to null (rather than a boolean set to false), then you can assertNotNull(exception) as well as assertTrue(exception instanceof ProcessException).

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java (hunk 4 above)
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Done.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java (hunk 5 above)
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: Ok. What I'm going to do there is make it so that empty query is only valid when there is no incoming connection.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java (hunk 4 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: I just realized that this block to check for an incoming connection and the Query property will get executed on each onTrigger. I believe it should be done during setup (with an OnScheduled method), check ExecuteSQL for an example. Otherwise this processor can generate an exception on each execution, and if the Run Schedule is zero seconds...

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java (hunk 5 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: I don't think you have access to connection information during validation (all you get is a ValidationContext), I was looking into adding it at one point, but never got around to it.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java (hunk 6 above)
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: Done

**Comment by MikeThomsen** [comment]
@mattyb149 Ready for review.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 7]
```diff
@@ -226,6 +242,8 @@ private ObjectWriter getObjectWriter(ObjectMapper mapper, String ppSetting) {
 
     @Override
     public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile input = session.get();
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: It's not checking for input == null here, that would be one of the requirements for the "flowfile-driven behavior" rather than the "timer-driven behavior". It might cause an NPE at line 266, and otherwise if null you may be treating it as flowfile-driven when really it is more like timer-driven. ExecuteSQL has a good example of how to handle this.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 8]
```diff
@@ -236,12 +254,33 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
                     context.getProperty(QUERY).evaluateAttributeExpressions().getValue());
         }
 
-        final Document query = context.getProperty(QUERY).isSet()
-                ? Document.parse(context.getProperty(QUERY).evaluateAttributeExpressions().getValue()) : null;
+        final Document query;
+        if (!context.hasIncomingConnection() && !context.getProperty(QUERY).isSet()) {
+            query = Document.parse("{}");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: I thought this was going to be a validation error? It can be done in OnScheduled, see ExecuteSQL for an example. Otherwise how would the user know that his/her configuration won't actually perform any work?

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java [hunk 9]
```diff
@@ -69,6 +70,7 @@
     @Before
     public void setup() {
         runner = TestRunners.newTestRunner(GetMongo.class);
+        //runner.setProperty(GetMongo.QUERY, "{}");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/GetMongoTest.java]: This can now be removed

**Comment by MikeThomsen** [comment]
@mattyb149 I broke the query location unit test down into three distinct cases:

* Verify that an empty query w/ no flowfile defaults to timer-driven behavior with `{}` as the query.
* Verify that a query param will work when set.
* Verify that it can read from the body w/ EL support.

**Comment by MikeThomsen** [comment]
Two of the 3 builds succeeded, and there are now five tests that cover the "read query" behavior so I think the tires have been pretty well kicked on this one.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 10]
```diff
@@ -236,12 +261,33 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
         }
 
-        final Document query = context.getProperty(QUERY).isSet()
-                ? Document.parse(context.getProperty(QUERY).evaluateAttributeExpressions().getValue()) : null;
+        final Document query;
+        if (context.getProperty(QUERY).isSet()) {
+            String queryStr = context.getProperty(QUERY).evaluateAttributeExpressions(input).getValue();
+            query = Document.parse(queryStr);
+        } else if (!context.getProperty(QUERY).isSet() && input == null) {
+            query = Document.parse("{}");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: The documentation for the Query property says that if there is no incoming connection and Query is not set, then it results in an error. But this like issues an empty query which appears to fetch all results.

Also please check line 234 above, looks like we need to do that later, after "queryStr" is populated for example. Currently the Query Attribute will not be populated if there is an incoming flow file.

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Ok. I think what I'll do is just put the query attribute on there no matter what if the query attribute is set because that'll help people figure out what happened in every scenario where a query was sent to Mongo.

File: nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java [hunk 11]
```diff
@@ -76,13 +77,28 @@
...
+    static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .name("original")
+            .description("All input flowfiles that are part of a successful query execution go here.")
             .build();
+
+    static final PropertyDescriptor QUERY = new PropertyDescriptor.Builder()
+        .name("Query")
+        .description("The selection criteria to do the lookup. If the field is left blank, it will look for input from" +
+                " an incoming connection from another processor to provide the query as a valid JSON document inside of " +
+                "the flowfile's body. If this field is left blank and a timer is enabled instead of an incoming connection, " +
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: This needs to be reverted to the original description of behavior when blank, that the entire collection will be queried. If this is not the intent, and an error should be thrown, then you'll need to change the logic around line 264

Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/GetMongo.java]: Done. Sorry about that.

**Comment by mattyb149** [comment]
+1 LGTM, ran unit tests and contrib-check, checked documentation and verified behavior on a NiFi instance, with and without incoming connections, Query parameter set, self-loops, etc.  Thanks for the improvement! Merging to master

**Comment by MikeThomsen** [comment]
Thanks. I'll rebase the other one and get started there.
